### PR TITLE
Make sure we support keyed services in client apps

### LINKED ISSUE #18992: Add Support for KeyedServices
### Description

`ServiceProvider` in Maui does not support KeyedServices

### Steps to Reproduce

`Microsoft.Maui.Controls.Application.Current?.Handler.MauiContext?.Services.GetRequiredKeyedService<T>(key);`

### Link to public reproduction project repository

_No response_

### Version with bug

8.0.3

### Is this a regression from previous behavior?

No, this is something new

### Last version that worked well

Unknown/Other

### Affected platforms

I was *not* able test on other platforms

### Affected platform versions

_No response_

### Did you find any workaround?

No

### Relevant log output

```shell
System.InvalidOperationException: 'This service provider doesn't support keyed services.'
```

### PR DESCRIPTION
**Pull request by mattleibow**
### Description of Change

New things in .NET 7 and we can't use it out the box.

Fixes #18992

### CODE REVIEW

**Comment by Eilon** [comment]
@BrennanConroy / @halter73 - could one of you have a look at adding this support to .NET MAUI?

**Comment by halter73** [comment]
I'm heading on vacation for the next two weeks. @benjaminpetit might also be able to help.

File: src/Core/src/MauiContext.cs [hunk 1]
```diff
@@ -72,6 +72,20 @@ public void AddSpecific(Type type, Func<object, object?> getter, object state)
...
 				_scopeStatic[type] = (state, getter);
 			}
+
+			public object? GetKeyedService(Type serviceType, object? serviceKey)
+			{
+				if (Inner is IKeyedServiceProvider provider)
+					return provider.GetKeyedService(serviceType, serviceKey);
+
+				// we know this won't work, but we need to call it to throw the right exception
+				return Inner.GetRequiredKeyedService(serviceType, serviceKey);
```
Comment (BrennanConroy) [inline src/Core/src/MauiContext.cs:82]: What do you mean throw the "right" exception? Doesn't this just throw an exception like "no service for serviceKey"? You want an error like https://github.com/dotnet/aspnetcore/blob/cc9cff31eb828f5849c07afc46b08baeda42b399/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L793

File: src/Core/src/MauiContext.cs [hunk 2]
```diff
@@ -72,6 +72,20 @@ public void AddSpecific(Type type, Func<object, object?> getter, object state)
 			{
 				_scopeStatic[type] = (state, getter);
 			}
+
+			public object? GetKeyedService(Type serviceType, object? serviceKey)
+			{
+				if (Inner is IKeyedServiceProvider provider)
+					return provider.GetKeyedService(serviceType, serviceKey);
```
Comment (BrennanConroy) [inline src/Core/src/MauiContext.cs:79]: No braces?

File: src/Core/tests/UnitTests/MauiContextTests.cs [hunk 3]
```diff
@@ -99,6 +99,23 @@ public void CloneCanOverrideIncludeService()
 			Assert.Same(obj2, second.Services.GetService<TestThing>());
 		}
 
+		[Fact]
+		public void MauiContextSupportsKeyedServices()
```
Comment (BrennanConroy) [inline src/Core/tests/UnitTests/MauiContextTests.cs:103]: Negative test?

**Review by BrennanConroy — COMMENTED** [review]
Does Maui look for attributes anywhere? There is the `FromKeyedServicesAttribute` if that's relevant to Maui.

File: src/Core/src/MauiContext.cs (hunk 2 above)
Comment (mattleibow) [reply src/Core/src/MauiContext.cs:79]: We don't always for single line statements - especially in short methods.

File: src/Core/tests/UnitTests/MauiContextTests.cs (hunk 3 above)
Comment (mattleibow) [reply src/Core/tests/UnitTests/MauiContextTests.cs:103]: All the other tests are the negative. We are not actually implementing a new service provider, but rather have a proxy/wrapper instance that allows for the ability to inject a value for a specific instance of a wrapper. This is needed for Android which has a weird way of doing things, but the wrapper just redirects all calls to the `Inner` instance.

**Comment by mattleibow** [comment]
Thanks for the review! I added a test for the attributes. All seems to be working. 

Since maui "service provider" is just a wrapper of the real provider with some extra buttons, no new logic was really written.

**Review by PureWeen — APPROVED** [review]

**Pull request merged**